File: tests/algorithms/x/OHP/strings/reverse_letters.php
<?php
error_reporting(E_ALL & ~E_DEPRECATED);
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
function _panic($msg) {
    fwrite(STDERR, strval($msg));
    exit(1);
}
$__start_mem = memory_get_usage();
$__start = _now();
  function mochi_split($s, $sep) {
  $res = [];
  $current = '';
  $i = 0;
  while ($i < strlen($s)) {
  $ch = substr($s, $i, $i + 1 - $i);
  if ($ch == $sep) {
  $res = _append($res, $current);
  $current = '';
} else {
  $current = $current . $ch;
}
  $i = $i + 1;
};
  $res = _append($res, $current);
  return $res;
};
  function join_with_space($xs) {
  $s = '';
  $i = 0;
  while ($i < count($xs)) {
  $s = $s . $xs[$i];
  if ($i + 1 < count($xs)) {
  $s = $s . ' ';
}
  $i = $i + 1;
};
  return $s;
};
  function reverse_str($s) {
  $res = '';
  $i = strlen($s) - 1;
  while ($i >= 0) {
  $res = $res . substr($s, $i, $i + 1 - $i);
  $i = $i - 1;
};
  return $res;
};
  function reverse_letters($sentence, $length) {
  $words = mochi_split($sentence, ' ');
  $result = [];
  $i = 0;
  while ($i < count($words)) {
  $word = $words[$i];
  if (strlen($word) > $length) {
  $result = _append($result, reverse_str($word));
} else {
  $result = _append($result, $word);
}
  $i = $i + 1;
};
  return join_with_space($result);
};
  function test_reverse_letters() {
  if (reverse_letters('Hey wollef sroirraw', 3) != 'Hey fellow warriors') {
  _panic('test1 failed');
}
  if (reverse_letters('nohtyP is nohtyP', 2) != 'Python is Python') {
  _panic('test2 failed');
}
  if (reverse_letters('1 12 123 1234 54321 654321', 0) != '1 21 321 4321 12345 123456') {
  _panic('test3 failed');
}
  if (reverse_letters('racecar', 0) != 'racecar') {
  _panic('test4 failed');
}
};
  function main() {
  test_reverse_letters();
  echo rtrim(reverse_letters('Hey wollef sroirraw', 3)), PHP_EOL;
};
  main();
$__end = _now();
$__end_mem = memory_get_peak_usage(true);
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/algorithms/transpiler/FS/strings/reverse_letters.fs
// Generated 2025-08-11 15:32 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec split (s: string) (sep: string) =
    let mutable __ret : string array = Unchecked.defaultof<string array>
    let mutable s = s
    let mutable sep = sep
    try
        let mutable res: string array = Array.empty<string>
        let mutable current: string = ""
        let mutable i: int = 0
        while i < (String.length (s)) do
            let ch: string = string (s.[i])
            if ch = sep then
                res <- Array.append res [|current|]
                current <- ""
            else
                current <- current + ch
            i <- i + 1
        res <- Array.append res [|current|]
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
and join_with_space (xs: string array) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable xs = xs
    try
        let mutable s: string = ""
        let mutable i: int = 0
        while i < (Seq.length (xs)) do
            s <- s + (_idx xs (int i))
            if (i + 1) < (Seq.length (xs)) then
                s <- s + " "
            i <- i + 1
        __ret <- s
        raise Return
        __ret
    with
        | Return -> __ret
and reverse_str (s: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable s = s
    try
        let mutable res: string = ""
        let mutable i: int = (String.length (s)) - 1
        while i >= 0 do
            res <- res + (string (s.[i]))
            i <- i - 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
and reverse_letters (sentence: string) (length: int) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable sentence = sentence
    let mutable length = length
    try
        let words: string array = split (sentence) (" ")
        let mutable result: string array = Array.empty<string>
        let mutable i: int = 0
        while i < (Seq.length (words)) do
            let word: string = _idx words (int i)
            if (String.length (word)) > length then
                result <- Array.append result [|(reverse_str (word))|]
            else
                result <- Array.append result [|word|]
            i <- i + 1
        __ret <- join_with_space (result)
        raise Return
        __ret
    with
        | Return -> __ret
and test_reverse_letters () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        if (reverse_letters ("Hey wollef sroirraw") (3)) <> "Hey fellow warriors" then
            failwith ("test1 failed")
        if (reverse_letters ("nohtyP is nohtyP") (2)) <> "Python is Python" then
            failwith ("test2 failed")
        if (reverse_letters ("1 12 123 1234 54321 654321") (0)) <> "1 21 321 4321 12345 123456" then
            failwith ("test3 failed")
        if (reverse_letters ("racecar") (0)) <> "racecar" then
            failwith ("test4 failed")
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        test_reverse_letters()
        printfn "%s" (reverse_letters ("Hey wollef sroirraw") (3))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()
